File: stores/traxsource.js
const fetch = require('node-fetch');
const cheerio = require('cheerio');

const STORE_NAME = 'Traxsource';
const STORE_URL = 'https://www.traxsource.com';

async function search(query) {
  try {
    const url = `https://www.traxsource.com/search?term=${encodeURIComponent(query)}`;
    const res = await fetch(url, {
      timeout: 10000,
      headers: {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
        'Accept': 'text/html,application/xhtml+xml',
        'Accept-Language': 'en-US,en;q=0.9',
      },
    });

    if (!res.ok) return [];

    const html = await res.text();
    const $ = cheerio.load(html);
    const results = [];

    $('.trk-row, .search-trk-row').each((_, el) => {
      const $el = $(el);
      const title = $el.find('.title a, .trk-name a').first().text().trim();
      const artist = $el.find('.artists a, .trk-artists a').map((_, a) => $(a).text().trim()).get().join(', ');
      const label = $el.find('.label a, .trk-label a').first().text().trim();
      const genre = $el.find('.genre a, .trk-genre a').first().text().trim();
      const price = $el.find('.add-cart .price, .buy-btn, .trk-price').first().text().trim();
      const link = $el.find('.title a, .trk-name a').first().attr('href');
      const img = $el.find('img.lazy, img.trk-art').first().attr('data-src') || $el.find('img').first().attr('src');

      if (title) {
        results.push({
          title,
          artist,
          label,
          genre,
          bpm: $el.find('.bpm, .trk-bpm').first().text().trim() || null,
          key: $el.find('.key, .trk-key').first().text().trim() || null,
          duration: $el.find('.duration, .trk-duration').first().text().trim(),
          price: price || '$1.49',
          priceValue: parsePrice(price || '$1.49'),
          currency: 'USD',
          artwork: img || null,
          url: link ? (link.startsWith('http') ? link : `${STORE_URL}${link}`) : `${STORE_URL}/search?term=${encodeURIComponent(query)}`,
          store: STORE_NAME,
          storeIcon: 'traxsource',
          releaseDate: '',
        });
      }
    });

    return results.slice(0, 25);
  } catch (err) {
    console.error(`Traxsource search error: ${err.message}`);
    return [];
  }
}

function parsePrice(str) {
  const match = str.match(/[\d.]+/);
  return match ? parseFloat(match[0]) : null;
}

module.exports = { search, STORE_NAME };

File: stores/traxsource.py
from __future__ import annotations

import re
from urllib.parse import quote

import requests
from bs4 import BeautifulSoup

from models import TrackResult

STORE_NAME = "Traxsource"
_STORE_URL = "https://www.traxsource.com"
_HEADERS = {
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
    "Accept": "text/html,application/xhtml+xml",
    "Accept-Language": "en-US,en;q=0.9",
}


def _parse_price(s: str) -> float | None:
    m = re.search(r"[\d.]+", s)
    return float(m.group()) if m else None


def search(query: str) -> list[TrackResult]:
    try:
        url = f"{_STORE_URL}/search?term={quote(query)}"
        resp = requests.get(url, headers=_HEADERS, timeout=10)
        if not resp.ok:
            return []

        soup = BeautifulSoup(resp.text, "html.parser")
        results: list[TrackResult] = []

        for el in soup.select(".trk-row, .search-trk-row"):
            title_el = el.select_one(".title a, .trk-name a")
            title = title_el.get_text(strip=True) if title_el else ""
            artist = ", ".join(a.get_text(strip=True) for a in el.select(".artists a, .trk-artists a"))
            label_el = el.select_one(".label a, .trk-label a")
            label = label_el.get_text(strip=True) if label_el else ""
            genre_el = el.select_one(".genre a, .trk-genre a")
            genre = genre_el.get_text(strip=True) if genre_el else ""
            price_el = el.select_one(".add-cart .price, .buy-btn, .trk-price")
            price_text = price_el.get_text(strip=True) if price_el else ""
            link = title_el["href"] if title_el and title_el.get("href") else None
            img_el = el.select_one("img.lazy, img.trk-art")
            img = (img_el.get("data-src") or img_el.get("src")) if img_el else None
            if not img:
                first_img = el.select_one("img")
                img = first_img.get("src") if first_img else None

            if title:
                bpm_el = el.select_one(".bpm, .trk-bpm")
                key_el = el.select_one(".key, .trk-key")
                dur_el = el.select_one(".duration, .trk-duration")

                full_url = link if link and link.startswith("http") else f"{_STORE_URL}{link}" if link else f"{_STORE_URL}/search?term={quote(query)}"
                results.append(TrackResult(
                    title=title,
                    artist=artist,
                    label=label,
                    genre=genre,
                    bpm=bpm_el.get_text(strip=True) if bpm_el else None,
                    key=key_el.get_text(strip=True) if key_el else None,
                    duration=dur_el.get_text(strip=True) if dur_el else "",
                    price=price_text or "$1.49",
                    price_value=_parse_price(price_text or "$1.49"),
                    currency="USD",
                    artwork=img,
                    url=full_url,
                    store=STORE_NAME,
                    store_icon="traxsource",
                    release_date="",
                ))

        return results[:25]
    except Exception as e:
        print(f"Traxsource search error: {e}")
        return []
